### PR TITLE
Avoid MooseEnum string comparisons

### LINKED ISSUE #5211: Plasticity uses string to MooseEnum comparisons in time critical places
@WilkAndy I can fix this. MooseEnums should be compared using their integer index values instead. This is very noticeable in a profiling run.

### PR DESCRIPTION
**Pull request by dschwen**
10% speedup.
Closes #5211

### CODE REVIEW

**Comment by dschwen** [comment]
Ping @wilkandy

**Comment by dschwen** [comment]
Hmm, that timeout failure seems to be real. I'll investigate.

**Comment by WilkAndy** [comment]
Hi Dan,

It may be because those tests change the deactivation scheme as the return-map algorithm proceeds.  I apologise for the very complicated nature of the code: the algorithm IS just super complicated.  But if you want to refactor, please feel free!  Crazy code...

a

**Comment by dschwen** [comment]
Ok, I'm pretty sure it is because I dropped a reference in ```changeScheme``` which modifies some passed in parameters. @wilkandy your functions have a _lot_ of parameters. Can we keep state in member variables instead? I know you have limited time, so I may look into changing this.

**Comment by WilkAndy** [comment]
I wrote the functions to have zillions of arguments deliberately because i wanted to ensure people knew which variables were changed by functions, and which variables remained the same.  Just because it's pretty tricky code in places.  Is there a better way to show future coders this information?

**Comment by dschwen** [comment]
The drawback is that you have a sh#t-ton of positional arguments, which make the code pretty hard to read on the caller side (and users are non the wiser as to which parameters are actually changed!).
Can we lump some parameters into state objects? for example ```stress, stress_old, intnl, intnl_old, plastic_strain, plastic_strain_old``` and possibly others? I guess refactoring this would take some time.

**Comment by dschwen** [comment]
This is ready to merge. But one question @WilkAndy it seems that some plasticity tests start with a residual of 0 (for the first and all successive timesteps, for example multi.rock1_heavy) and do no solve whatsoever. What is the deal with that?

**Comment by WilkAndy** [comment]
Probably most of the tests are designed to start at residual=0 and do no solve.  They are testing the return-map algorithm, which is really the (super) tricky thing here.  Often i've used a nonzero initial_stress, with all displacements fixed to zero , and then the algorithm should return the stress to the yield surface.  Then i check that the returned stress is the desired value.

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C [hunk 1]
```diff
@@ -673,33 +709,56 @@ ComputeMultiPlasticityStress::canAddConstraints(const std::vector<bool> & act, c
 }
 
 bool
-ComputeMultiPlasticityStress::canChangeScheme(const MooseEnum & current_deactivation_scheme, const bool & can_revert_to_dumb)
+ComputeMultiPlasticityStress::canChangeScheme(DeactivationSchemeEnum current_deactivation_scheme,
+                                              const bool & can_revert_to_dumb)
```
Comment (permcody) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: No big deal, but a pointer or reference is bigger than a simple type like a bool. Usually you don't need to pass built in types as references. It doesn't save you anything and might actually cost you more if the compiler doesn't optimize that away.

**Pull request merged**